opennebula/resource_opennebula_user.go

```diff
@@ -0,0 +1,264 @@
+				Optional:    true,
+				Default:     0,
+				Description: "Primary (Default) Group ID of the user. Defaults to 0",
+			},
+			"groups": {
+				Type:        schema.TypeList,
+				Optional:    true,
+				Description: "List of group IDs to add to the user",
+				Elem: &schema.Schema{
+					Type: schema.TypeInt,
+				},
+			},
+			"quotas": quotasSchema(),
+		},
+	}
+}
+
+func getUserController(d *schema.ResourceData, meta interface{}) (*goca.UserController, error) {
+	controller := meta.(*goca.Controller)
+	var uc *goca.UserController
+
+	// Try to find the User by ID, if specified
+	if d.Id() != "" {
+		uid, err := strconv.ParseUint(d.Id(), 10, 64)
+		if err != nil {
+			return nil, err
+		}
+		uc = controller.User(int(uid))
+	}
+
+	// Otherwise, try to find the User by name as the de facto compound primary key
+	if d.Id() == "" {
+		uid, err := controller.Users().ByName(d.Get("name").(string))
+		if err != nil {
+			return nil, err
+		}
+		uc = controller.User(uid)
+	}
+
+	return uc, nil
+}
+
+func resourceOpennebulaUserCreate(d *schema.ResourceData, meta interface{}) error {
+	controller := meta.(*goca.Controller)
+
+	user_name := d.Get("name").(string)
+	user_auth_driver := d.Get("auth_driver").(string)
+	var user_password string
+	if user_auth_driver != "ldap" {
+		user_password_interface, ok := d.GetOk("password")
+		if !ok {
+			return fmt.Errorf("Password cannot be empty if auth_driver is: %s", user_auth_driver)
+		}
+		user_password = user_password_interface.(string)
+	}
+	user_groups_list := d.Get("groups").([]interface{})
+	user_groups := make([]int, 0)
+
+	// Start Group array with Primary group
+	user_groups = append(user_groups, d.Get("primary_group").(int))
+
+	// add groups to user if list provided
+	for _, gid := range user_groups_list {
+		user_groups = append(user_groups, gid.(int))
+	}
+
+	userID, err := controller.Users().Create(user_name, user_password, user_auth_driver, user_groups)
+	if err != nil {
+		return err
+	}
+	d.SetId(fmt.Sprintf("%v", userID))
+
+	uc := controller.User(userID)
+
+	if _, ok := d.GetOk("quotas"); ok {
+		err = uc.Quota(generateQuotas(d))
+		if err != nil {
+			return err
+		}
+	}
+
+	return resourceOpennebulaUserRead(d, meta)
+}
+
+func resourceOpennebulaUserRead(d *schema.ResourceData, meta interface{}) error {
+	uc, err := getUserController(d, meta)
+	if err != nil {
+		if NoExists(err) {
+			log.Printf("[WARN] Removing user %s from state because it no longer exists in", d.Get("name"))
+			d.SetId("")
+			return nil
+		}
+		return err
+	}
+
+	// TODO: fix it after 5.10 release
+	// Force the "decrypt" bool to false to keep ONE 5.8 behavior
+	user, err := uc.Info(false)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(strconv.FormatUint(uint64(user.ID), 10))
+	d.Set("name", user.Name)
+	sum := sha256.Sum256([]byte(d.Get("password").(string)))
+	if fmt.Sprintf("%x", sum) == user.Password {
+		d.Set("password", d.Get("password").(string))
+	} else {
+		return fmt.Errorf("Unexpected Password")
```

Comment:
could be slightly more explicit like: `password doesn't match`, or `password changed`

Reply:
indeed